wave_matching/include/wave/matching/impl/multi_matcher_impl.hpp

```diff
@@ -0,0 +1,100 @@
+#ifndef WAVE_MULTI_MATCHER_IMPL_HPP
+#define WAVE_MULTI_MATCHER_IMPL_HPP
+
+#include <iostream>
+
+namespace wave {
+
+template <class T, class R>
+MultiMatcher<T, R>::~MultiMatcher() {
+    // Destroy all the workers
+    this->stop = true;
+    this->ip_condition.notify_all();
+    for (int id = 0; id < this->n_thread; ++id) {
+        this->pool.at(id).join();
+    }
+}
+
+template <class T, class R>
+void MultiMatcher<T, R>::initPool(std::string path) {
+    this->config = path;
+    for (int i = 0; i < this->n_thread; i++) {
+        if (this->config.empty()) {
+            this->matchers.push_back(T(R()));
+        } else {
+            this->matchers.push_back(T(R(this->config)));
+        }
+        this->pool.push_back(std::thread(&MultiMatcher<T, R>::spin, this, i));
```

Comment:
Why does it matter that the MultiMatcher class stores N matcher objects and N threads? That was my question! :wink:

You should declare each variable as locally as possible. That is for readability and maintainability. 

As a corollary: if the variables are declared non-locally in some roundabout way, it suggests to readers that your design _requires_ it for some reason. But I can't tell what that is.